controller/setting_controller.go

```diff
@@ -223,10 +223,44 @@ func (sc *SettingController) syncSetting(key string) (err error) {
 		if err := sc.syncUpgradeChecker(); err != nil {
 			return err
 		}
+		return nil
 	case string(types.SettingNameBackupTarget), string(types.SettingNameBackupTargetCredentialSecret), string(types.SettingNameBackupstorePollInterval):
 		if err := sc.syncBackupTarget(); err != nil {
 			return err
 		}
+		return nil
+	case string(types.SettingNameKubernetesClusterAutoscalerEnabled):
+		if err := sc.updateKubernetesClusterAutoscalerEnabled(); err != nil {
+			return err
+		}
+		return nil
+	case string(types.SettingNameSupportBundleFailedHistoryLimit):
+		if err := sc.cleanupFailedSupportBundles(); err != nil {
+			return err
+		}
+		return nil
+	case string(types.SettingNameLogLevel):
+		if err := sc.updateLogLevel(); err != nil {
+			return err
+		}
+		return nil
+	case string(types.SettingNameV2DataEngine):
+		if err := sc.updateV2DataEngine(); err != nil {
+			return err
+		}
+		return nil
+	}
+
+	// SettingCategoryDangerZone and they will try to delete pods to update the settings.
+	volumesDetached, err := sc.ds.AreAllVolumesDetached()
+	if err != nil {
+		return errors.Wrapf(err, "failed to check volume detachment for %v setting update", name)
+	}
+	if !volumesDetached {
+		return &types.ErrorInvalidState{Reason: fmt.Sprintf("failed to apply %v setting to Longhorn workloads when there are attached volumes", name)}
```

Comment:
Don't throw error in this case, because it will be just requeued three times.

Have a debug log as well. 

[/controller/setting_controller.go#L194-L209](https://github.com/longhorn/longhorn-manager/blob/b517601153190d762c88e38442e8037ff6b94391/controller/setting_controller.go#L194-L209)
```go
func (sc *SettingController) handleErr(err error, key interface{}) {
	if err == nil {
		sc.queue.Forget(key)
		return
	}

	log := sc.logger.WithField("Setting", key)
	if sc.queue.NumRequeues(key) < maxRetries {
		handleReconcileErrorLogging(log, err, "Failed to sync Longhorn setting")
		sc.queue.AddRateLimited(key)
		return
	}

	utilruntime.HandleError(err)
	handleReconcileErrorLogging(log, err, "Dropping Longhorn setting out of the queue")
	sc.queue.Forget(key)
```

Instead, do AddRateLimited directly, similar to below. It will be eventually updated, but it could be a delay to take effect depending on the rate, even though all volumes have been detached by users. Therefore, we need to update the description of impacted settings to explain this eventual effect behavior. Also, the doc needs to be updated.

[/controller/node_controller.go#L580-L588](https://github.com/longhorn/longhorn-manager/blob/b517601153190d762c88e38442e8037ff6b94391/controller/node_controller.go#L580-L588)
```go
func (nc *NodeController) enqueueNodeRateLimited(obj interface{}) {
	key, err := controller.KeyFunc(obj)
	if err != nil {
		utilruntime.HandleError(fmt.Errorf("couldn't get key for object %#v: %v", obj, err))
		return
	}

	nc.queue.AddRateLimited(key)
}
```

Reply:
If we call the function e.g. `enqueueSettingRateLimited` to `AddRateLimited` and return `nil` back to `handleErr`.

```golang
func (sc *SettingController) handleErr(err error, key interface{}) {
	if err == nil {
		sc.queue.Forget(key)
		return
	}
```
It will forget the key, and another controller will get the item and then enqueue by `AddRateLimited` but it is not delayed actually.

Modified to get the error message and enqueue it by `sc.queue.AddRateLimited(key)` in `handleErr`.

Reply:
>Did you mean the code in [/controller/node_controller.go#L580-L588](https://github.com/longhorn/longhorn-manager/blob/b517601153190d762c88e38442e8037ff6b94391/controller/node_controller.go#L580-L588) not working as expected?

It will retry to reconcile the node CR (test the replica eviction when updating replicas fails), but it will not delay the item enqueued as expected.
That's what I found when testing the codes by adding `enqueueSettingRateLimited' with  `queue.AddRateLimited(key)`.

Reply:
> It will retry to reconcile the node CR (test the replica eviction when updating replicas fails), but it will not delay the item enqueued as expected.
> 

To be clear, didn't u mean the item enqueued Will immediately be forgotten during handling the nil error in the last reply?

What "delay" u meant? 

> That's what I found when testing the codes for retrying the Setting CR by adding `enqueueSettingRateLimited` with `queue.AddRateLimited(key)`.

So u found it on as requeued immediately but without rate limitation?

Reply:
>To be clear, didn't u mean the item enqueued immediately be forgotten (forget) during handling the nil error in the last reply?

Yes, IIUC, it will be forgotten if no other controller starts to handle the event.

>What "delay" did u mean?
So u found the event can be requeued immediately but without rate limitation?

Yes, and "delay" means `the event can be requeued immediately but without rate limitation`

Reply:
Removed that returning errors to retry updating danger zone settings when there are volumes attached.